### PR TITLE
Explain how to prepare a Go 1.5 development environment

### PR DESCRIPTION
**Pull request by gdm85**
Other minor documentation fixes

### CODE REVIEW

**Comment by nathanleclaire** [comment]
Hm, @dmp42 has some changes in the pipe for this doc as well which I will probably merge first, but I'm interested, so stay tuned.

**Comment by dmp42** [comment]
@gdm85 do you think you can rebase now the other PRs have been merged? Thanks a lot.

**Comment by gdm85** [comment]
@dmp42 done.

Please note that I am omitting the link to Go 1.5 installation on purpose (I don't think it fits the scope of this documentation, but it's arguable) and then providing an example about how to work on sources directly under a ``GOPATH``. I am not sure if this is a best practice or not, but I couldn't figure out a better way to reuse the project-provided dependencies.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -14,29 +14,41 @@ Docker does things, you'll feel right at home.
...
+    $ make build
 
-## Building
+## Local Go 1.5 development environment
+
+Make sure the source code directory is under a correct directory structure to use Go 1.5 vendoring;
+example of cloning and preparing the correct environment ``GOPATH``:
+````
+    mkdir -p projects/docker-machine/src/github.com/docker
+    export GOPATH="$PWD/projects/docker-machine"
```
Comment (dmp42) [inline CONTRIBUTING.md]: you probably want `GOPATH="$PWD/projects/docker-machine":$GOPATH` here

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -14,29 +14,41 @@ Docker does things, you'll feel right at home.
...
+    cd projects/docker-machine/src/github.com/docker
+    git clone https://github.com/docker/machine
+    cd machine
+    export GOPATH="$GOPATH:$PWD"
```
Comment (dmp42) [inline CONTRIBUTING.md]: I believe this one is useless.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -14,29 +14,41 @@ Docker does things, you'll feel right at home.
...
-[Install and setup go](https://golang.org/doc/install), then clone the machine repository inside your gopath.
+    $ export USE_CONTAINER=true
+    $ make build
 
-## Building
+## Local Go 1.5 development environment
+
+Make sure the source code directory is under a correct directory structure to use Go 1.5 vendoring;
+example of cloning and preparing the correct environment ``GOPATH``:
+````
```
Comment (dmp42) [inline CONTRIBUTING.md]: Only three backticks here.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -14,29 +14,41 @@ Docker does things, you'll feel right at home.
...
-1. A running instance of Docker
+1. A running instance of Docker or a Golang 1.5 development environment
 2. The `bash` shell
 3. [Make](https://www.gnu.org/software/make/)
 
-Call `export USE_CONTAINER=true` to instruct the build system to use containers to build.
+## Build using Docker containers
 
-# Alternative: build using go only
+To build the ``docker-machine`` binary using containers, simply run:
```
Comment (dmp42) [inline CONTRIBUTING.md]: You just need one backtick for these.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -14,29 +14,41 @@ Docker does things, you'll feel right at home.
...
 
-To build the docker-machine binary, simply run:
+At this point, simply run:
 
     $ make build
 
-From the Machine repository's root. You will now have a `bin/docker-machine`.
+## Built binary
+
+After the build is complete a ``bin/docker-machine`` binary will be created.
```
Comment (dmp42) [inline CONTRIBUTING.md]: One backtick.

**Comment by dmp42** [comment]
Thanks a lot @gdm85 !
I added a couple of remarks.
Also, we should still separate the native golang approach versus the docker only approach (people going with the latter *don't* need golang at all).

File: CONTRIBUTING.md (hunk 2 above)
Comment (gdm85) [inline CONTRIBUTING.md]: Well, the imports from this same project (`docker/machine`) are not vendor'ed (obviously) and not in `GOPATH`, I remember trying without this line and getting errors.

**Comment by gdm85** [comment]
@dmp42 I am still learning markdown, apparently :)

As per the approaches split, do you think it should be made visible in a different way?

**Comment by gdm85** [comment]
I've removed also the extra GOPATH, tested, it works

**Comment by jeanlaurent** [comment]
We could probably add the default simplest way of getting ready to code on machine which is @dgageot setup, this is what we want to achieve with this guide, make it easy to contribute.

And add an optional setup, where you don't share you `GOPATH` with other projects ( as @gdm85 suggested ). 

As my personal experience, I did them both, and there is no doubt the former is easier.

**Comment by jeanlaurent** [comment]
@dmp42 IMHO this could be merged "as is", as it brings value right now. We should decide of the easiest way of getting up to speed on the project in another dedicated PR.

**Comment by gdm85** [comment]
@jeanlaurent my personal opinion is that `GOPATH` is something that should be replaced by a command-line option (as it works in proven compilers like gcc/g++ with `-I`), but that is off-topic here :)

What I wanted to contribute here is a way for even an **occasional** contributor to do the right thing; I find myself contributing after weeks or a month and not updating the packages in `GOPATH` is dangerous in such cases.

With my latest commit 8ac92c2 I simplified the process even further by making use of `go get`; I didn't squash it to leave the option to you guys to either use it or drop it.

**Comment by gdm85** [comment]
@dgageot ops, added the missing `go get` now :)

**Comment by jeanlaurent** [comment]
thx @gdm85

@dmp42 ready to review & merge.

**Comment by dmp42** [comment]
LGTM

@jeanlaurent you can hit that merge button at will.

**Comment by nathanleclaire** [comment]
Should `GOPATH` really point to the Docker Machine repo's location on disk?  I'd generally be inclined towards recommending `export GOPATH="$(pwd)/vendor:$GOPATH"` or, better yet, `export GO15VENDOREXPERIMENT=1`.

**Comment by dmp42** [comment]
`GO15VENDOREXPERIMENT=1` will not help (will just make golang resolve `vendor` but won't add the current directory to GOPATH)

Exporting `GOPATH="$(pwd)/vendor:$GOPATH"` is usually the right approach in a proper golang setup where you work on multiple different stuff - but it might also create difficulties for newcomers wrt package resolution, and is not needed per-se if you just want to hack on machine.

Either way, a follow-up PR can certainly enhance this if this is a problem.

**Comment by gdm85** [comment]
@nathanleclaire now that you mention it, I notice that my wording is misleading as I wrote `to use Go 1.5 vendoring` although this project doesn't use it yet; sorry for that.

In this PR I saw the approach more from the perspective of allowing contributions without having to fetch sub-packages in a separate `GOPATH`, thus as easy and self-contained as possible.

**Comment by nathanleclaire** [comment]
@gdm85 No problem, looks like a good start for now, and I'll revisit and PR any nitpicks I have.

**Pull request merged**